### PR TITLE
fix: possible to have same slug for team and user in an organization 9985 cal 2122 

### LINKED ISSUE #9985: [CAL-2122] Possible to have same slug for team and user in an organization
I am able to create a user in organization with same username as a team slug of that organization

GitHub-hosted image

[Source image](https://uploads.linear.app/e86bf957-d82f-465e-b205-135559f4b623/0953e31b-1ad9-49be-bc56-6f89cb878b54/3652f769-d2ad-4b44-9f44-57be667c2ef2)

[Source image](https://github.com/calcom/cal.com/assets/1780212/3c34c598-f7e7-48e6-a6dc-57c47b18793d)

It is important to not allow that because in an organization a team's booking page is available as [acme.cal.com/sales](http://acme.cal.com/sales) and a user with username 'sales' is also available at [acme.cal.com/sales](http://acme.cal.com/sales)

Due to this [acme.cal.com/sales/secret](http://acme.cal.com/sales/secret) (an automatically created event is giving 404)

[Source image](https://github.com/calcom/cal.com/assets/1780212/2d6aab1a-f726-40d2-b758-c1a28243f151)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2122](https://linear.app/calcom/issue/CAL-2122/possible-to-have-same-slug-for-team-and-user-in-an-organization)</sub>

### PR DESCRIPTION
**Pull request by alannnc**
## What does this PR do?

- Moves invite token validation before user creation
- If token has teamId we collect every team.slug in the org or just use that team.slug so a username cannot have same team.slug.

Fixes #9985 

## Requirement/Documentation

  - If there is a requirement document, please, share it here.
  - If there is ab UI/UX design document, please, share it here.

## Type of change

  - [X] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

  - Are there environment variables that should be set? No
  - What are the minimal test data to have? A newly created orgs with a couple teams already created
  - What is expected (happy path) to have (input and output)? When inviting a user via link and he ends up on signup page, the username is validated against team slugs in the entire org or just that simple team. If username has the same name as the team he should see an error that says Username is taken.
  - Any other important info that could help to test that PR

## Mandatory Tasks

  - [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: apps/web/pages/api/auth/signup.ts [hunk 1]
```diff
@@ -70,6 +71,35 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
+
+    if (dayjs(foundToken?.expires).isBefore(dayjs())) {
+      return res.status(401).json({ message: "Token expired" });
+    }
+    if (foundToken?.teamId) {
+      const isValidUsername = await validateUsernameInOrg(username, foundToken?.teamId);
+
+      if (!isValidUsername) {
+        return res.status(409).json({ message: "Username already taken" });
+      }
```
Comment (alannnc) [inline apps/web/pages/api/auth/signup.ts:99]: This code was running before user creation I think we can save a little if moved up here.

File: apps/web/pages/api/auth/signup.ts [hunk 2]
```diff
@@ -70,6 +71,35 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
+
+    if (!foundToken) {
+      return res.status(401).json({ message: "Invalid Token" });
+    }
+
+    if (dayjs(foundToken?.expires).isBefore(dayjs())) {
+      return res.status(401).json({ message: "Token expired" });
+    }
+    if (foundToken?.teamId) {
+      const isValidUsername = await validateUsernameInOrg(username, foundToken?.teamId);
```
Comment (alannnc) [inline apps/web/pages/api/auth/signup.ts:95]: Additionally running this function created on this PR.

File: apps/web/pages/api/auth/signup.ts [hunk 3]
```diff
@@ -88,91 +118,75 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
+              parentId: team.parentId,
             },
-          });
-        }
+            accepted: false,
+          },
+          data: {
+            accepted: true,
+          },
+        });
```
Comment (alannnc) [inline apps/web/pages/api/auth/signup.ts:189]: Removed a unnecessary conditional so all of this code moved up.

File: packages/lib/validateUsernameInOrg.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import prisma from "@calcom/prisma";
+
+import slugify from "./slugify";
+
+export const validateUsernameInOrg = async (usernameSlug: string, teamId: number): Promise<boolean> => {
+  try {
+    let takenSlugs = [];
+    const teamsFound = await prisma.team.findMany({
+      where: {
+        OR: [{ id: teamId }, { parentId: teamId }],
```
Comment (alannnc) [inline packages/lib/validateUsernameInOrg.ts:10]: Find every team that has parentId as teamId or just a single team.

Comment (alannnc) [reply packages/lib/validateUsernameInOrg.ts:10]: TODO: What if org has only one team?

File: packages/trpc/server/routers/viewer/teams/inviteMember/inviteMember.handler.ts [hunk 5]
```diff
@@ -86,7 +86,7 @@ export const inviteMemberHandler = async ({ ctx, input }: InviteMemberOptions) =
           isCalcomMember: true,
         };
         /**
-         * Here we want to redirect to a differnt place if onboarding has been completed or not. This prevents the flash of going to teams -> Then to onboarding - also show a differnt email template.
+         * Here we want to redirect to a different place if onboarding has been completed or not. This prevents the flash of going to teams -> Then to onboarding - also show a different email template.
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/teams/inviteMember/inviteMember.handler.ts:89]: typos

**Review by leog — CHANGES_REQUESTED** [review]
Changes look good. I think we could also take care of updating username as well, which I think relies on the domain and that’s not accurate, as the app works under “app” not the org subdomain.

**Comment by alannnc** [comment]
> Changes look good. I think we could also take care of updating username as well, which I think relies on the domain and that’s not accurate, as the app works under “app” not the org subdomain.

Can we do name update on other PR? @leog

**Comment by leog** [comment]
> Can we do name update on other PR? @leog

Of course @alannnc makes sense. Approving.

**Review by leog — APPROVED** [review]
LGTM, follow-up PR to tackle updating username is expected.

**Comment by alannnc** [comment]
> LGTM, follow-up PR to tackle updating username is expected.
Here you go:
https://github.com/calcom/cal.com/issues/10196
@leog

**Pull request merged**